File: ios/Packages/FatwaBotKit/Sources/CoreKit/AnalyticsTracking.swift
import Foundation

/// Product-analytics boundary. Lives in CoreKit (mirrors `ActivityEventRecording`
/// and `HapticsProviding`) so feature modules depend on the protocol, never on a
/// transport or a vendor SDK, and stay testable with `NoopAnalyticsTracking`.
///
/// On iOS this is backed by our OWN ingest (`POST /v1/analytics/events`) rather
/// than a third-party SDK — see docs/features/analytics-and-crash-reporting.md.
///
/// ## What must never be sent
///
/// This is a worship app used by anonymous users. Treat the following as
/// forbidden payloads, not merely discouraged (the server rejects them too):
///
///  - **Search queries.** "Is X permissible" is among the most sensitive things a
///    user types. Report *that* a search happened, never *what* it said.
///  - **Location.** Not coordinates, not city. Prayer times are computed
///    on-device (ADR-0003) precisely so location never leaves it.
///  - **Identity.** No display name, email, user id, or push token.
///  - **Content bodies.** Du'a/hadith text or personal wird notes. A stable id or
///    category key is fine; free text is not.
///
/// Screen names, stable category keys and counts are the intended payloads.
public protocol AnalyticsTracking: Sendable {
    /// A screen became visible. `screen` must be a stable, non-PII key.
    func screenView(_ screen: String)

    /// A product event. `params` values must be stable keys or numbers — never
    /// user-authored text (see above).
    func event(_ name: String, params: [String: String])

    /// A handled error, for triage. Only the error's *type* is reported — never
    /// its message, which can carry user data.
    func nonFatal(_ error: Error)
}

public extension AnalyticsTracking {
    func event(_ name: String) { event(name, params: [:]) }
}

public struct NoopAnalyticsTracking: AnalyticsTracking {
    public init() {}
    public func screenView(_ screen: String) {}
    public func event(_ name: String, params: [String: String]) {}
    public func nonFatal(_ error: Error) {}
}

/// Canonical event + screen names, so the same concept isn't reported as
/// `dua_opened` here and `open_dua` there — which silently splits a metric
/// across two names and stays invisible until someone questions the numbers.
/// Kept identical to Android's `AnalyticsEvents` so one dashboard serves both.
public enum AnalyticsEvents {
    // Screens
    public static let screenHome = "home"
    public static let screenWorship = "worship"
    public static let screenSettings = "settings"
    public static let screenPrayer = "prayer_times"
    public static let screenQibla = "qibla"
    public static let screenTasbeeh = "tasbeeh"
    public static let screenAzkar = "azkar"
    public static let screenDua = "dua"
    public static let screenAwrad = "awrad"
    public static let screenHadith = "hadith"
    public static let screenJourney = "journey"

    // Events
    public static let screenView = "screen_view"
    public static let widgetOpenedApp = "widget_opened_app"
    public static let searchSubmitted = "search_submitted"
    public static let nonFatalError = "non_fatal_error"
    public static let tasbeehSessionCompleted = "tasbeeh_session_completed"
    public static let azkarCategoryCompleted = "azkar_category_completed"
    public static let notificationPrefsChanged = "notification_prefs_changed"

    // Param keys
    public static let paramScreen = "screen"
    public static let paramRoute = "route"
    public static let paramCategory = "category"
    public static let paramCount = "count"
    public static let paramErrorType = "error_type"
}

File: android/core/common/src/main/kotlin/com/fatwabot/core/common/AnalyticsTracking.kt
package com.fatwabot.core.common

/**
 * Product-analytics boundary. Hoisted here (mirrors [ActivityEventRecording] and
 * HapticsProviding) so feature modules never depend on Firebase directly — the
 * SDK stays an `:app`-only implementation detail and features stay unit-testable
 * with [NoopAnalyticsTracking].
 *
 * Android **dual-sends**: the bound implementation is a composite over
 * `FirebaseAnalyticsTracker` (free aggregate/retention dashboards, and
 * Crashlytics uses it for breadcrumbs) and
 * `com.fatwabot.core.network.BackendAnalyticsRecorder`, which posts to our OWN
 * ingest so the first-party dataset covers both platforms. iOS single-sends to
 * that same ingest (no Firebase SDK there by design), so the event and param
 * names below must stay byte-identical to iOS `AnalyticsEvents` — a mismatch
 * silently splits a metric across two names in one dashboard.
 *
 * ## What must never be logged
 *
 * This is a worship app used by anonymous users, and analytics goes to a third
 * party. Treat the following as forbidden payloads, not merely discouraged:
 *
 *  - **Search queries.** "Is X permissible" is among the most sensitive things a
 *    user types. Log that a search happened, never what it said.
 *  - **Location.** Not coordinates, not city. Prayer times are computed on-device
 *    precisely so location never leaves it.
 *  - **Identity.** No display name, email, user id, or push token.
 *  - **Content bodies.** Du'a/hadith text or personal wird notes. An id or a
 *    category key is fine; free text is not.
 *
 * Counts, screen names, and stable category keys are the intended payloads.
 */
interface AnalyticsTracking {
    /** A screen became visible. [screen] must be a stable, non-PII key. */
    fun screenView(screen: String)

    /**
     * A product event. [params] values must be stable keys or numbers — never
     * user-authored text (see the class doc).
     */
    fun event(name: String, params: Map<String, String> = emptyMap())

    /**
     * Record a handled error for triage without crashing. Pass an exception, not
     * a formatted user-facing string, so it groups correctly.
     */
    fun nonFatal(error: Throwable)

    /** Mirrors the user's privacy choice into the underlying SDKs. */
    fun setCollectionEnabled(enabled: Boolean)
}

/** Default binding for tests and any build without a reporting backend. */
class NoopAnalyticsTracking : AnalyticsTracking {
    override fun screenView(screen: String) {}
    override fun event(name: String, params: Map<String, String>) {}
    override fun nonFatal(error: Throwable) {}
    override fun setCollectionEnabled(enabled: Boolean) {}
}

/**
 * Canonical event + screen names, so the same concept isn't logged as
 * `dua_opened` on one screen and `open_dua` on another — which silently splits
 * a metric across two names in the dashboard.
 */
object AnalyticsEvents {
    // Screens
    const val SCREEN_HOME = "home"
    const val SCREEN_WORSHIP = "worship"
    const val SCREEN_SETTINGS = "settings"
    const val SCREEN_PRAYER = "prayer_times"
    const val SCREEN_QIBLA = "qibla"
    const val SCREEN_TASBEEH = "tasbeeh"
    const val SCREEN_AZKAR = "azkar"
    const val SCREEN_DUA = "dua"
    const val SCREEN_AWRAD = "awrad"
    const val SCREEN_HADITH = "hadith"
    const val SCREEN_JOURNEY = "journey"

    // Events
    const val SCREEN_VIEW = "screen_view"
    const val WIDGET_OPENED_APP = "widget_opened_app"
    const val SEARCH_SUBMITTED = "search_submitted"
    const val NON_FATAL_ERROR = "non_fatal_error"
    const val TASBEEH_SESSION_COMPLETED = "tasbeeh_session_completed"
    const val AZKAR_CATEGORY_COMPLETED = "azkar_category_completed"
    const val NOTIFICATION_PREFS_CHANGED = "notification_prefs_changed"

    // Param keys
    const val PARAM_SCREEN = "screen"
    const val PARAM_ROUTE = "route"
    const val PARAM_CATEGORY = "category"
    const val PARAM_COUNT = "count"
    const val PARAM_ERROR_TYPE = "error_type"
}
